File: Trials/Network/RUDP/ReliableUDPListener.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace RUDP
{
    public class ReliableUDPListener : ReliableUDP
    {
        AutoResetEvent AcceptWaitEvent = new AutoResetEvent(false);

        /// <summary>
        /// コネクションされたクライアントのリスト.
        /// </summary>
        SafeList<ReliableUDP> AcceptedClientList = new SafeList<ReliableUDP>();

        /// <summary>
        /// 受信したクライアントのリスト.
        /// </summary>
        Dictionary<IPEndPoint, ReliableUDPClient> RecievedClientMap = new Dictionary<IPEndPoint, ReliableUDPClient>();

        Thread RecieveThread;

        public ReliableUDPListener(UdpClient udp)
        {
            this.Udp = udp;

            this.RecieveThread = new Thread(ProcRecieve);
            this.RecieveThread.Start();
        }

        public ReliableUDP Accept()
        {
            if (this.IsClosed)
            {
                //throw new 
            }

            if (this.AcceptedClientList.Count == 0)
            {
                this.AcceptWaitEvent.WaitOne();
            }

            var ret = this.AcceptedClientList[this.AcceptedClientList.Count - 1];
            this.AcceptedClientList.Remove(ret);

            return ret;
        }

        private void ProcRecieve()
        {
            while (!this.WillClose)
            {
                IPEndPoint remoteEp = new IPEndPoint(IPAddress.Any, 0);
                var segment = base.OnRecieveSegment(ref remoteEp);

                if (segment != null)
                {
                    ReliableUDPClient client = null;

                    lock (this.RecievedClientMap)
                    {
                        if (!this.IsClosed)
                        {
                            if (segment is SynchronousSegment)
                            {
                                // まずは登録済みかどうか探す.
                                if (!this.RecievedClientMap.ContainsKey(remoteEp))
                                {
                                    // 未登録の場合は登録する.

                                    client = ReliableUDPClient.Create(this, remoteEp);
                                    this.RecievedClientMap.Add(remoteEp, client);
                                }
                            }
                        }

                        if (client == null)
                        {
                            client = this.RecievedClientMap[remoteEp];
                        }

                        if (client != null)
                        {
                            client.EnqueueSegment(segment);
                        }
                    }
                }
            }
        }

        class ReliableUDPClient : ReliableUDP
        {
            IPEndPoint RemoteEp;

            ReliableUDPListener Parent;

            SafeQueue<Segment> RecievedSegmentQueue = new SafeQueue<Segment>();

            AutoResetEvent QueueWaitEvent = new AutoResetEvent(false);

            static public ReliableUDPClient Create(
                ReliableUDPListener server,
                IPEndPoint ep)
            {
                var ret = new ReliableUDPClient(server, ep);
                return ret;
            }

            private ReliableUDPClient(ReliableUDPListener server, IPEndPoint ep)
                : base(server.Udp, new Parameter())
            {
                this.RemoteEp = ep;
                this.Parent = server;
            }

            public void EnqueueSegment(Segment segment)
            {
                if (!this.IsClosed)
                {
                    this.RecievedSegmentQueue.Enqueue(segment);
                    this.QueueWaitEvent.Set();
                }
            }

            protected override void OnOpenConnection()
            {
#if false
                // 一定サイズを超えていたいら
                if (this.Parent.AcceptedClientList.Count > MAX)
                {
                    // Wait
                }
#endif

                this.Parent.AcceptedClientList.Add(this);

                this.Parent.AcceptWaitEvent.Set();
            }

            protected override Segment OnRecieveSegment()
            {
                if (!this.IsClosed)
                {
                    if (this.RecievedSegmentQueue.Count == 0)
                    {
                        this.QueueWaitEvent.WaitOne();
                    }

                    this.QueueWaitEvent.Reset();

                    return this.RecievedSegmentQueue.Dequeue();
                }

                return null;
            }

            protected override void OnSendSegment(Segment segment)
            {
                Console.WriteLine("Send Segment [{0}]", segment.ToString());

                using (var ws = new MemoryStream())
                {
                    segment.WriteBytes(ws);
                    var bytes = ws.GetBuffer();
                    var length = ws.Position;
                    this.Udp.Send(bytes, (int)length, this.RemoteEp);
                }
            }

            // TODO
            // close
        }
    }
}

File: Libs/source/network/rudp/ReliableUDPListener.cpp
#include "network/rudp/ReliableUDPListener.h"
#include "network/rudp/segment/Segment.h"
#include "network/NetworkUDP_libuv.h"
#include "network/IPv4Endpoint.h"

namespace izanagi {
namespace net {
    void ReliableUDPListener::Init(
        IMemoryAllocator* allocator,
        Udp* udp)
    {
        if (m_Udp == nullptr) {
            m_allocator = allocator;
            m_Udp = udp;

            m_RecieveThread = std::thread(std::bind(&ReliableUDPListener::ProcRecieve, this));
        }
    }

    ReliableUDP* ReliableUDPListener::Accept()
    {
        if (IsClosed())
        {
            IZ_ASSERT(IZ_FALSE);
            return nullptr;
        }

        std::lock_guard<std::mutex> locker(m_AcceptedClientListWaiter);

        if (m_AcceptedClientList.GetItemNum() == 0)
        {
            return nullptr;
        }

        auto item = m_AcceptedClientList.GetTop();
        auto ret = item->GetData();
        
        item->Leave();

        return ret;
    }

    void ReliableUDPListener::ProcRecieve()
    {
        while (!WillClose())
        {
            IPv4Endpoint remoteEp;
            auto segment = OnRecieveSegment(remoteEp);

            if (segment)
            {
                ReliableUDPClient* client = nullptr;

                if (!IsClosed())
                {
                    if (segment->IsSynchronousSegment())
                    {
                        // まずは登録済みかどうか探す.
                        if (!m_RecievedClientMap.ContainsKey(remoteEp))
                        {
                            // 未登録の場合は登録する.

                            client = ReliableUDPClient::Create(
                                m_allocator,
                                this,
                                remoteEp);

                            m_RecievedClientMap.Add(client->GetHashItem());
                        }
                    }
                }

                if (client == nullptr)
                {
                    client = m_RecievedClientMap.FindData(remoteEp);
                }

                if (client != nullptr)
                {
                    client->EnqueueSegment(segment);
                }
            }
        }
    }

    Udp* ReliableUDPListener::GetUdp()
    {
        return m_Udp;
    }

    ReliableUDPListener::ReliableUDPClient* ReliableUDPListener::ReliableUDPClient::Create(
        IMemoryAllocator* allocator,
        ReliableUDPListener* server,
        const IPv4Endpoint& ep)
    {
        void* p = ALLOC(allocator, sizeof(ReliableUDPClient));
        VRETURN(p != IZ_NULL);

        ReliableUDPClient* ret = new(p)ReliableUDPClient(server, ep);
        ret->Init(allocator, server->GetUdp(), RUDPParameter());

        return ret;
    }

    ReliableUDPListener::ReliableUDPClient::ReliableUDPClient(
        ReliableUDPListener* server,
        const IPv4Endpoint& ep)
    {
        m_RemoteEp = ep;
        m_Parent = server;

        m_listItem.Init(this);
        m_hashItem.Init(ep, this);
    }

    void ReliableUDPListener::ReliableUDPClient::EnqueueSegment(Segment* segment)
    {
        if (!IsClosed())
        {
            sys::Lock locker(*this);
            m_RecievedSegmentQueue.AddTail(segment->GetListItem());
        }
    }

    void ReliableUDPListener::ReliableUDPClient::OnOpenConnection()
    {
        std::lock_guard<std::mutex> locker(m_Parent->m_AcceptedClientListWaiter);

#if 0
        // 一定サイズを超えていたいら
        if (this.Parent.AcceptedClientList.Count > MAX)
        {
            // Wait
        }
#endif

        m_Parent->m_AcceptedClientList.AddTail(this->GetListItem());
    }

    Segment* ReliableUDPListener::ReliableUDPClient::OnRecieveSegment()
    {
        if (!IsClosed())
        {
            sys::Lock locker(*this);

            if (m_RecievedSegmentQueue.GetItemNum() == 0)
            {
                return nullptr;
            }

            auto item = m_RecievedSegmentQueue.GetTop();

            auto ret = item->GetData();

            item->Leave();

            return ret;
        }

        return nullptr;
    }

    void ReliableUDPListener::ReliableUDPClient::OnSendSegment(Segment* segment)
    {
        IZ_CHAR str[64];
        segment->ToString(str, sizeof(str));

        IZ_PRINTF("Send Segment [%s]\n", str);

        IZ_UINT8 tmp[64];

        void* data = tmp;
        IZ_UINT length = sizeof(tmp);

        if (m_isAllocated) {
            data = m_sendData;
            length = m_Parameter.MaxSegmentSize;
        }

        CMemoryOutputStream ws(data, length);

        segment->WriteBytes(&ws);
        auto bytes = ws.GetBuffer();
        auto size = ws.GetCurPos();
        m_Udp->sendTo(bytes, size, m_RemoteEp);
    }
}   // namespace net
}   // namespace izanagi
